Regression in Jetty 12.1.0: @OnWebSocketFrame and @OnWebSocketMessage can no longer coexist

**Jetty version(s)**
Jetty version: 12.1.0
Previous version (working): 12.0.25

**Jetty Environment**
ee10

**Java version/vendor** `(use: java -version)`
java21

**OS type/version**
Linux

### Description

After upgrading from **Jetty 12.0.25 → 12.1.0**, our WebSocket adapter class that combines both `@OnWebSocketMessage` and `@OnWebSocketFrame` no longer works.

In 12.0.25 this was valid and allowed us to process **text/binary messages** via `@OnWebSocketMessage` and handle **Pong frames** via `@OnWebSocketFrame`.
Starting with 12.1.0, the same code fails with `InvalidWebSocketException`.

---

### Reproducer

```java
@WebSocket
public class XWebSocketAdapter {
    @OnWebSocketMessage
    public void onWebSocketText(String text) {
        // handle text
    }

    @OnWebSocketMessage
    public void onWebSocketBinary(ByteBuffer data, @Out Callback out) {
        // handle binary
        out.succeed();
    }

    @OnWebSocketFrame
    public void onWebSocketFrame(Frame frame, @Out Callback out) {
        // specifically handle Pong frames
        if (frame.getOpCode() == OpCode.PONG) {
            // custom pong handling
        }
        out.succeed();
    }
}
```

---

### Actual Behavior (12.1.0)

```
Caused by: org.eclipse.jetty.websocket.api.exceptions.InvalidWebSocketException: Cannot replace previously assigned [FRAME Handler] at MethodHandle(XWebSocketAdapter,Session,Frame,Callback)void with public void cn.nextop.gadget.web.support.socket.XWebSocketAdapter.onWebSocketText(java.lang.String)
```

---

### Expected Behavior (same as 12.0.25)

* `@OnWebSocketMessage` should handle normal text/binary messages.
* `@OnWebSocketFrame` should be allowed to process frames (in our case, Pong).
* No conflict should occur.

---

### Questions

* Is this restriction in Jetty 12.1.0 intentional?

* If yes, what is the recommended migration path for applications that previously used `@OnWebSocketFrame` only for handling Pong frames while still using `@OnWebSocketMessage`?

* Jetty does not provide `@OnWebSocketPing` / `@OnWebSocketPong` annotations, so it is unclear what the intended replacement is for these cases.

---

### Additional Notes

* This is a **behavior change** between **12.0.25 (works)** and **12.1.0 (fails)**.
* If intentional, it would be helpful to clarify this in the release notes or migration guide.

---

--- a/jetty-core/jetty-websocket/jetty-websocket-jetty-api/src/main/java/org/eclipse/jetty/websocket/api/annotations/OnWebSocketPing.java
+++ b/jetty-core/jetty-websocket/jetty-websocket-jetty-api/src/main/java/org/eclipse/jetty/websocket/api/annotations/OnWebSocketPing.java
@@ -1,0 +1,35 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.websocket.api.annotations;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * <p>Annotation for methods to receive WebSocket ping events.</p>
+ * <p>Acceptable method patterns:</p>
+ * <ol>
+ * <li>{@code public void <methodName>(ByteBuffer payload)}</li>
+ * <li>{@code public void <methodName>(Session session, ByteBuffer payload)}</li>
+ * </ol>
+ */
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD)
+public @interface OnWebSocketPing
+{
+}

--- a/jetty-core/jetty-websocket/jetty-websocket-jetty-api/src/main/java/org/eclipse/jetty/websocket/api/annotations/OnWebSocketPing.java
+++ b/jetty-core/jetty-websocket/jetty-websocket-jetty-api/src/main/java/org/eclipse/jetty/websocket/api/annotations/OnWebSocketPing.java
@@ -20,7 +20,7 @@ import java.lang.annotation.RetentionPolicy;
 import java.lang.annotation.Target;
 
 /**
- * <p>Annotation for methods to receive WebSocket ping events.</p>
+ * <p>Annotation for methods to receive WebSocket PING events.</p>
  * <p>Acceptable method patterns:</p>
  * <ol>
  * <li>{@code public void <methodName>(ByteBuffer payload)}</li>

--- a/jetty-core/jetty-websocket/jetty-websocket-jetty-api/src/main/java/org/eclipse/jetty/websocket/api/annotations/OnWebSocketPong.java
+++ b/jetty-core/jetty-websocket/jetty-websocket-jetty-api/src/main/java/org/eclipse/jetty/websocket/api/annotations/OnWebSocketPong.java
@@ -1,0 +1,35 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.websocket.api.annotations;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * <p>Annotation for methods to receive WebSocket ping events.</p>
+ * <p>Acceptable method patterns:</p>
+ * <ol>
+ * <li>{@code public void <methodName>(ByteBuffer payload)}</li>
+ * <li>{@code public void <methodName>(Session session, ByteBuffer payload)}</li>
+ * </ol>
+ */
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD)
+public @interface OnWebSocketPong
+{
+}

--- a/jetty-core/jetty-websocket/jetty-websocket-jetty-api/src/main/java/org/eclipse/jetty/websocket/api/annotations/OnWebSocketPong.java
+++ b/jetty-core/jetty-websocket/jetty-websocket-jetty-api/src/main/java/org/eclipse/jetty/websocket/api/annotations/OnWebSocketPong.java
@@ -20,7 +20,7 @@ import java.lang.annotation.RetentionPolicy;
 import java.lang.annotation.Target;
 
 /**
- * <p>Annotation for methods to receive WebSocket ping events.</p>
+ * <p>Annotation for methods to receive WebSocket PONG events.</p>
  * <p>Acceptable method patterns:</p>
  * <ol>
  * <li>{@code public void <methodName>(ByteBuffer payload)}</li>

--- a/jetty-core/jetty-websocket/jetty-websocket-jetty-common/src/main/java/org/eclipse/jetty/websocket/common/JettyWebSocketFrameHandlerFactory.java
+++ b/jetty-core/jetty-websocket/jetty-websocket-jetty-common/src/main/java/org/eclipse/jetty/websocket/common/JettyWebSocketFrameHandlerFactory.java
@@ -35,6 +35,8 @@ import org.eclipse.jetty.websocket.api.annotations.OnWebSocketError;
 import org.eclipse.jetty.websocket.api.annotations.OnWebSocketFrame;
 import org.eclipse.jetty.websocket.api.annotations.OnWebSocketMessage;
 import org.eclipse.jetty.websocket.api.annotations.OnWebSocketOpen;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketPing;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketPong;
 import org.eclipse.jetty.websocket.api.annotations.WebSocket;
 import org.eclipse.jetty.websocket.api.exceptions.InvalidWebSocketException;
 import org.eclipse.jetty.websocket.common.internal.ByteBufferMessageSink;
@@ -304,6 +306,28 @@ public class JettyWebSocketFrameHandlerFactory extends ContainerLifeCycle
             final InvokerUtils.Arg CALLBACK = new InvokerUtils.Arg(Callback.class).required();
             MethodHandle methodHandle = InvokerUtils.mutatedInvoker(lookup, endpointClass, onmethod, SESSION, FRAME, CALLBACK);
             metadata.setFrameHandle(methodHandle, onmethod);
+        }
+
+        // OnWebSocketPing [0..1]
+        onmethod = ReflectUtils.findAnnotatedMethod(endpointClass, OnWebSocketPing.class);
+        if (onmethod != null)
+        {
+            assertSignatureValid(endpointClass, onmethod, OnWebSocketPing.class);
+            final InvokerUtils.Arg SESSION = new InvokerUtils.Arg(Session.class);
+            final InvokerUtils.Arg BUFFER = new InvokerUtils.Arg(ByteBuffer.class).required();
+            MethodHandle methodHandle = InvokerUtils.mutatedInvoker(lookup, endpointClass, onmethod, SESSION, BUFFER);
+            metadata.setPingHandle(methodHandle, onmethod);
+        }
+
+        // OnWebSocketPong [0..1]
+        onmethod = ReflectUtils.findAnnotatedMethod(endpointClass, OnWebSocketPong.class);
+        if (onmethod != null)
+        {
+            assertSignatureValid(endpointClass, onmethod, OnWebSocketPong.class);
+            final InvokerUtils.Arg SESSION = new InvokerUtils.Arg(Session.class);
+            final InvokerUtils.Arg BUFFER = new InvokerUtils.Arg(ByteBuffer.class).required();
+            MethodHandle methodHandle = InvokerUtils.mutatedInvoker(lookup, endpointClass, onmethod, SESSION, BUFFER);
+            metadata.setPongHandle(methodHandle, onmethod);
         }
 
         // OnWebSocketMessage [0..2]

--- a/jetty-core/jetty-websocket/jetty-websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/EchoSocket.java
+++ b/jetty-core/jetty-websocket/jetty-websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/EchoSocket.java
@@ -23,16 +23,16 @@ import org.eclipse.jetty.websocket.api.annotations.WebSocket;
 public class EchoSocket extends EventSocket
 {
     @Override
-    public void onMessage(String message) throws IOException
+    public void onTextMessage(String message) throws IOException
     {
-        super.onMessage(message);
+        super.onTextMessage(message);
         session.sendText(message, Callback.NOOP);
     }
 
     @Override
-    public void onMessage(ByteBuffer message, Callback callback) throws IOException
+    public void onBinaryMessage(ByteBuffer message, Callback callback) throws IOException
     {
-        super.onMessage(message, Callback.NOOP);
+        super.onBinaryMessage(message, Callback.NOOP);
         session.sendBinary(message, callback);
     }
 }

--- a/jetty-core/jetty-websocket/jetty-websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/ErrorCloseTest.java
+++ b/jetty-core/jetty-websocket/jetty-websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/ErrorCloseTest.java
@@ -99,9 +99,9 @@ public class ErrorCloseTest
         }
 
         @Override
-        public void onMessage(String message) throws IOException
+        public void onTextMessage(String message) throws IOException
         {
-            super.onMessage(message);
+            super.onTextMessage(message);
             if (methodsToThrow.contains("onMessage"))
                 throw new RuntimeException("throwing from onMessage");
         }

--- a/jetty-core/jetty-websocket/jetty-websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/EventSocket.java
+++ b/jetty-core/jetty-websocket/jetty-websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/EventSocket.java
@@ -83,6 +83,7 @@ public class EventSocket
         if (LOG.isDebugEnabled())
             LOG.debug("{}  onPingMessage(): {}", this, payload);
         pingMessages.add(BufferUtil.copy(payload));
+        session.sendPong(payload, Callback.NOOP);
     }
 
     @OnWebSocketPong

--- a/jetty-core/jetty-websocket/jetty-websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/EventSocket.java
+++ b/jetty-core/jetty-websocket/jetty-websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/EventSocket.java
@@ -28,6 +28,8 @@ import org.eclipse.jetty.websocket.api.annotations.OnWebSocketClose;
 import org.eclipse.jetty.websocket.api.annotations.OnWebSocketError;
 import org.eclipse.jetty.websocket.api.annotations.OnWebSocketMessage;
 import org.eclipse.jetty.websocket.api.annotations.OnWebSocketOpen;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketPing;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketPong;
 import org.eclipse.jetty.websocket.api.annotations.WebSocket;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -37,17 +39,17 @@ public class EventSocket
 {
     private static final Logger LOG = LoggerFactory.getLogger(EventSocket.class);
 
+    public final BlockingQueue<String> textMessages = new BlockingArrayQueue<>();
+    public final BlockingQueue<ByteBuffer> binaryMessages = new BlockingArrayQueue<>();
+    public final BlockingQueue<ByteBuffer> pongMessages = new BlockingArrayQueue<>();
+    public final BlockingQueue<ByteBuffer> pingMessages = new BlockingArrayQueue<>();
+    public final CountDownLatch openLatch = new CountDownLatch(1);
+    public final CountDownLatch errorLatch = new CountDownLatch(1);
+    public final CountDownLatch closeLatch = new CountDownLatch(1);
     public Session session;
-
-    public BlockingQueue<String> textMessages = new BlockingArrayQueue<>();
-    public BlockingQueue<ByteBuffer> binaryMessages = new BlockingArrayQueue<>();
-    public volatile int closeCode = StatusCode.UNDEFINED;
-    public volatile String closeReason;
-    public volatile Throwable error = null;
-
-    public CountDownLatch openLatch = new CountDownLatch(1);
-    public CountDownLatch errorLatch = new CountDownLatch(1);
-    public CountDownLatch closeLatch = new CountDownLatch(1);
+    public int closeCode = StatusCode.UNDEFINED;
+    public String closeReason;
+    public Throwable error = null;
 
     @OnWebSocketOpen
     public void onOpen(Session session)
@@ -59,20 +61,45 @@ public class EventSocket
     }
 
     @OnWebSocketMessage
-    public void onMessage(String message) throws IOException
+    public void onTextMessage(String message) throws IOException
     {
         if (LOG.isDebugEnabled())
-            LOG.debug("{}  onMessage(): {}", this, message);
-        textMessages.offer(message);
+            LOG.debug("{}  onTextMessage(): {}", this, message);
+        textMessages.add(message);
     }
 
     @OnWebSocketMessage
-    public void onMessage(ByteBuffer message, Callback callback) throws IOException
+    public void onBinaryMessage(ByteBuffer message, Callback callback) throws IOException
     {
         if (LOG.isDebugEnabled())
-            LOG.debug("{}  onMessage(): {}", this, message);
-        binaryMessages.offer(BufferUtil.copy(message));
+            LOG.debug("{}  onBinaryMessage(): {}", this, message);
+        binaryMessages.add(BufferUtil.copy(message));
         callback.succeed();
+    }
+
+    @OnWebSocketPing
+    public void onPingMessage(ByteBuffer payload)
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("{}  onPingMessage(): {}", this, payload);
+        pingMessages.add(BufferUtil.copy(payload));
+    }
+
+    @OnWebSocketPong
+    public void onPongMessage(ByteBuffer payload)
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("{}  onPongMessage(): {}", this, payload);
+        pongMessages.add(BufferUtil.copy(payload));
+    }
+
+    @OnWebSocketError
+    public void onError(Throwable cause)
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("{}  onError()", this, cause);
+        error = cause;
+        errorLatch.countDown();
     }
 
     @OnWebSocketClose
@@ -83,15 +110,6 @@ public class EventSocket
         this.closeCode = statusCode;
         this.closeReason = reason;
         closeLatch.countDown();
-    }
-
-    @OnWebSocketError
-    public void onError(Throwable cause)
-    {
-        if (LOG.isDebugEnabled())
-            LOG.debug("{}  onError(): {}", this, cause);
-        error = cause;
-        errorLatch.countDown();
     }
 
     @Override

--- a/jetty-core/jetty-websocket/jetty-websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/ExplicitDemandTest.java
+++ b/jetty-core/jetty-websocket/jetty-websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/ExplicitDemandTest.java
@@ -61,9 +61,9 @@ public class ExplicitDemandTest
         }
 
         @Override
-        public void onMessage(String message) throws IOException
+        public void onTextMessage(String message) throws IOException
         {
-            super.onMessage(message);
+            super.onTextMessage(message);
             if (!"suspend".equals(message))
                 session.demand();
         }

--- a/jetty-core/jetty-websocket/jetty-websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/PingPongTest.java
+++ b/jetty-core/jetty-websocket/jetty-websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/PingPongTest.java
@@ -1,0 +1,187 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.websocket.tests;
+
+import java.net.URI;
+import java.nio.ByteBuffer;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.ServerConnector;
+import org.eclipse.jetty.util.BlockingArrayQueue;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.websocket.api.Callback;
+import org.eclipse.jetty.websocket.api.Session;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketClose;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketPing;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketPong;
+import org.eclipse.jetty.websocket.api.annotations.WebSocket;
+import org.eclipse.jetty.websocket.client.WebSocketClient;
+import org.eclipse.jetty.websocket.server.WebSocketUpgradeHandler;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class PingPongTest
+{
+    Server _server;
+    ServerConnector _connector;
+    WebSocketClient _client;
+    AnnotatedPingPongSocket _annotatedPingPongSocket;
+    ListenerPingPongSocket _listenerPingPongSocket;
+
+    @BeforeEach
+    public void before() throws Exception
+    {
+        _server = new Server();
+        _connector = new ServerConnector(_server);
+        _server.addConnector(_connector);
+        _client = new WebSocketClient();
+
+        WebSocketUpgradeHandler upgradeHandler = WebSocketUpgradeHandler.from(_server);
+        _annotatedPingPongSocket = new AnnotatedPingPongSocket();
+        _listenerPingPongSocket = new ListenerPingPongSocket();
+        upgradeHandler.getServerWebSocketContainer().addMapping("/annotatedEndpoint", (req, resp, cb) -> _annotatedPingPongSocket);
+        upgradeHandler.getServerWebSocketContainer().addMapping("/listenerEndpoint", (req, resp, cb) -> _listenerPingPongSocket);
+        _server.setHandler(upgradeHandler);
+
+        _server.start();
+        _client.start();
+    }
+
+    @AfterEach
+    public void after() throws Exception
+    {
+        _client.stop();
+        _server.stop();
+    }
+
+    @Test
+    public void testAnnotatedPingPong() throws Exception
+    {
+        URI uri = URI.create("ws://localhost:" + _connector.getLocalPort() + "/annotatedEndpoint");
+        EventSocket clientEndpoint = new EventSocket();
+        Session session = _client.connect(clientEndpoint, uri).get();
+
+        // Test sending a ping and receiving it on the server.
+        session.sendPing(BufferUtil.toBuffer("ping from test"), Callback.NOOP);
+        ByteBuffer pingPayload = _annotatedPingPongSocket.pings.poll(5, TimeUnit.SECONDS);
+        assertThat(BufferUtil.toString(pingPayload), equalTo("ping from test"));
+
+        // Test sending a pong and receiving it on the server.
+        session.sendPong(BufferUtil.toBuffer("pong from test"), Callback.NOOP);
+        ByteBuffer pongPayload = _annotatedPingPongSocket.pongs.poll(5, TimeUnit.SECONDS);
+        assertThat(BufferUtil.toString(pongPayload), equalTo("pong from test"));
+
+        // Close the session.
+        session.close();
+        assertTrue(_annotatedPingPongSocket.closed.await(5, TimeUnit.SECONDS));
+        assertTrue(clientEndpoint.closeLatch.await(5, TimeUnit.SECONDS));
+
+        // Verify that no more pings or pongs were sent or received.
+        assertTrue(_annotatedPingPongSocket.pings.isEmpty());
+        assertTrue(_annotatedPingPongSocket.pongs.isEmpty());
+        assertTrue(clientEndpoint.pongMessages.isEmpty());
+        assertTrue(clientEndpoint.pingMessages.isEmpty());
+    }
+
+    @Test
+    public void testListenerPingPong() throws Exception
+    {
+        URI uri = URI.create("ws://localhost:" + _connector.getLocalPort() + "/listenerEndpoint");
+        EventSocket clientEndpoint = new EventSocket();
+        Session session = _client.connect(clientEndpoint, uri).get();
+
+        // Test sending a ping and receiving it on the server.
+        session.sendPing(BufferUtil.toBuffer("ping from test"), Callback.NOOP);
+        ByteBuffer pingPayload = _listenerPingPongSocket.pings.poll(5, TimeUnit.SECONDS);
+        assertThat(BufferUtil.toString(pingPayload), equalTo("ping from test"));
+
+        // Test sending a pong and receiving it on the server.
+        session.sendPong(BufferUtil.toBuffer("pong from test"), Callback.NOOP);
+        ByteBuffer pongPayload = _listenerPingPongSocket.pongs.poll(5, TimeUnit.SECONDS);
+        assertThat(BufferUtil.toString(pongPayload), equalTo("pong from test"));
+
+        // Close the session.
+        session.close();
+        assertTrue(_listenerPingPongSocket.closed.await(5, TimeUnit.SECONDS));
+        assertTrue(clientEndpoint.closeLatch.await(5, TimeUnit.SECONDS));
+
+        // Verify that no more pings or pongs were sent or received.
+        assertTrue(_listenerPingPongSocket.pings.isEmpty());
+        assertTrue(_listenerPingPongSocket.pongs.isEmpty());
+        assertTrue(clientEndpoint.pongMessages.isEmpty());
+        assertTrue(clientEndpoint.pingMessages.isEmpty());
+    }
+
+    @WebSocket
+    public static class AnnotatedPingPongSocket
+    {
+        private final BlockingQueue<ByteBuffer> pings = new BlockingArrayQueue<>();
+        private final BlockingQueue<ByteBuffer> pongs = new BlockingArrayQueue<>();
+        private final CountDownLatch closed = new CountDownLatch(1);
+
+        @OnWebSocketPing
+        public void onPing(Session session, ByteBuffer payload)
+        {
+            assertNotNull(session);
+            pings.add(payload);
+        }
+
+        @OnWebSocketPong
+        public void onPong(Session session, ByteBuffer payload)
+        {
+            assertNotNull(session);
+            pongs.add(payload);
+        }
+
+        @OnWebSocketClose
+        public void onClose(int statusCode, String reason)
+        {
+            closed.countDown();
+        }
+    }
+
+    public static class ListenerPingPongSocket implements Session.Listener.AutoDemanding
+    {
+        private final BlockingQueue<ByteBuffer> pings = new BlockingArrayQueue<>();
+        private final BlockingQueue<ByteBuffer> pongs = new BlockingArrayQueue<>();
+        private final CountDownLatch closed = new CountDownLatch(1);
+
+        @Override
+        public void onWebSocketPing(ByteBuffer payload)
+        {
+            pings.add(payload);
+        }
+
+        @Override
+        public void onWebSocketPong(ByteBuffer payload)
+        {
+            pongs.add(payload);
+        }
+
+        @Override
+        public void onWebSocketClose(int statusCode, String reason)
+        {
+            closed.countDown();
+        }
+    }
+}

--- a/jetty-core/jetty-websocket/jetty-websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/proxy/WebSocketProxyTest.java
+++ b/jetty-core/jetty-websocket/jetty-websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/proxy/WebSocketProxyTest.java
@@ -422,9 +422,9 @@ public class WebSocketProxyTest
     public static class OnTextThrowingSocket extends EventSocket
     {
         @Override
-        public void onMessage(String message) throws IOException
+        public void onTextMessage(String message) throws IOException
         {
-            super.onMessage(message);
+            super.onTextMessage(message);
             throw new IllegalStateException("simulated onMessage error");
         }
     }

